dev/devicelab/lib/framework/runner.dart

```diff
@@ -160,17 +163,55 @@ Future<VMIsolateRef> _connectToRunnerIsolate(Uri vmServiceUri) async {
       await (await WebSocket.connect(url)).close();
 
       // Look up the isolate.
-      final VMServiceClient client = VMServiceClient.connect(url);
-      final VM vm = await client.getVM();
-      final VMIsolateRef isolate = vm.isolates.single;
-      final String response = await isolate.invokeExtension('ext.cocoonRunnerReady') as String;
-      if (response != 'ready')
+      final VmService client = await vmServiceConnectUri(url);
+      VM vm = await client.getVM();
+      while (vm.isolates.isEmpty) {
+        await Future<void>.delayed(const Duration(seconds: 1));
+        vm = await client.getVM();
+      }
+      final IsolateRef isolate = vm.isolates.first;
+      final Response response = await client.callServiceExtension('ext.cocoonRunnerReady', isolateId: isolate.id);
+      if (response.json['response'] != 'ready')
         throw 'not ready yet';
-      return isolate;
+      return ConnectionResult(client, isolate);
     } catch (error) {
       if (stopwatch.elapsed > const Duration(seconds: 10))
         print('VM service still not ready after ${stopwatch.elapsed}: $error\nContinuing to retry...');
       await Future<void>.delayed(const Duration(milliseconds: 50));
     }
   }
 }
+
+class ConnectionResult {
+  ConnectionResult(this.vmService, this.isolate);
+
+  final VmService vmService;
+  final IsolateRef isolate;
+}
+
+/// The cocoon client sends an invalid VM service response, we need to intercept it.
+Future<VmService> vmServiceConnectUri(String wsUri, {Log log}) async {
+  final WebSocket socket = await WebSocket.connect(wsUri);
+  final StreamController<dynamic> controller = StreamController<dynamic>();
+  final Completer<dynamic> streamClosedCompleter = Completer<dynamic>();
+  socket.listen(
+    (dynamic data) {
+      final Map<String, dynamic> rawData = json.decode(data as String) as Map<String, dynamic> ;
```

Comment:
This is where the fun begins

Reply:
Consider adding an `onError` callback to the `listen` call that will propagate a useful stack trace.